packages/cli/test/unit/util/deploy/create-git-meta.test.ts

```diff
@@ -0,0 +1,147 @@
+    const configPath = join(fixture('no-email'), 'config');
+    const data = await getRepoData(configPath);
+    expect(data?.provider).toEqual('github');
+    expect(data?.org).toEqual('MatthewStanciu');
+    expect(data?.repo).toEqual('git-test');
+  });
+  it('gets repo data for no-origin', async () => {
+    const configPath = join(fixture('no-origin'), 'config');
+    const data = await getRepoData(configPath);
+    expect(data).toBeUndefined();
+  });
+});
+
+describe('parseRepoUrl', () => {
+  it('should parse github https url', () => {
+    const parsedUrl = parseRepoUrl('https://github.com/vercel/vercel.git');
+    expect(parsedUrl?.provider).toEqual('github');
+    expect(parsedUrl?.org).toEqual('vercel');
+    expect(parsedUrl?.repo).toEqual('vercel');
+  });
+  it('should parse github git url', () => {
+    const parsedUrl = parseRepoUrl('git://github.com/vercel/vercel.git');
+    expect(parsedUrl?.provider).toEqual('github');
+    expect(parsedUrl?.org).toEqual('vercel');
+    expect(parsedUrl?.repo).toEqual('vercel');
+  });
+  it('should parse github ssh url', () => {
+    const parsedUrl = parseRepoUrl('git@github.com:vercel/vercel.git');
+    expect(parsedUrl?.provider).toEqual('github');
+    expect(parsedUrl?.org).toEqual('vercel');
+    expect(parsedUrl?.repo).toEqual('vercel');
+  });
+
+  it('should parse gitlab https url', () => {
+    const parsedUrl = parseRepoUrl(
+      'https://gitlab.com/gitlab-examples/knative-kotlin-app.git'
+    );
+    expect(parsedUrl?.provider).toEqual('gitlab');
+    expect(parsedUrl?.org).toEqual('gitlab-examples');
+    expect(parsedUrl?.repo).toEqual('knative-kotlin-app');
+  });
+  it('should parse gitlab ssh url', () => {
+    const parsedUrl = parseRepoUrl(
+      'git@gitlab.com:gitlab-examples/knative-kotlin-app.git'
+    );
+    expect(parsedUrl?.provider).toEqual('gitlab');
+    expect(parsedUrl?.org).toEqual('gitlab-examples');
+    expect(parsedUrl?.repo).toEqual('knative-kotlin-app');
+  });
+
+  it('should parse bitbucket https url', () => {
+    const parsedUrl = parseRepoUrl(
+      'https://bitbucket.org/atlassianlabs/maven-project-example.git'
+    );
+    expect(parsedUrl?.provider).toEqual('bitbucket');
+    expect(parsedUrl?.org).toEqual('atlassianlabs');
+    expect(parsedUrl?.repo).toEqual('maven-project-example');
+  });
+  it('should parse bitbucket ssh url', () => {
+    const parsedUrl = parseRepoUrl(
+      'git@bitbucket.org:atlassianlabs/maven-project-example.git'
+    );
+    expect(parsedUrl?.provider).toEqual('bitbucket');
+    expect(parsedUrl?.org).toEqual('atlassianlabs');
+    expect(parsedUrl?.repo).toEqual('maven-project-example');
+  });
+});
+
+describe('createGitMeta', () => {
+  it('gets git metata from test-github', async () => {
+    const directory = fixture('test-github');
+    try {
+      await fs.rename(join(directory, 'git'), join(directory, '.git'));
+      const data = (await createGitMeta(directory)) as GitHubMeta;
```

Comment:
Every time we use `as`, we're telling Typescript to trust that we know the true type of something. This will often be true, but we don't have the automated protection of it being true.

In this case, we could check for the type we expect and fail the test if it's something else. This will provide a clearer error message if the test fails. Typescript itself doesn't let you do actual type checking at runtime, but you can do it indirectly.

One way I like to do this in cases like these is to add a property I can use to distinguish each type, like so:

```
export interface GitHubMeta {
  type: 'github';
  // ...
}

export interface GitLabMeta {
  type: 'gitlab';
  // ...
}

export interface BitbucketMeta {
  type: 'bitbucket';
  // ...
}
```

Then we can say:

```
await fs.rename(join(directory, 'git'), join(directory, '.git'));
const data = await createGitMeta(directory);
if (data.type !== 'github') {
  throw new Error(`Expected createGitMeta to return type "github', but received ${data.type}.`);
}
```

Typescript will assume the value of `data` after that block to be `GithubMeta`.

Reply:
```suggestion
      const data = await createGitMeta(directory);
      assert('githubCommitMessage' in data);
```